libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -97,45 +97,27 @@ proc rebalanceMesh(g: GossipSub, topic: string) {.async.} =
     if topic notin g.mesh:
       g.mesh[topic] = initHashSet[string]()
 
-    if g.mesh.getOrDefault(topic).len < GossipSubDlo:
-      trace "replenishing mesh", topic
-      # replenish the mesh if we're below GossipSubDlo
-      while g.mesh.getOrDefault(topic).len < GossipSubD:
-        trace "gathering peers", peers = g.mesh.getOrDefault(topic).len
-        await sleepAsync(1.millis) # don't starve the event loop
-        var id: string
-        if topic in g.fanout and g.fanout.getOrDefault(topic).len > 0:
-          trace "getting peer from fanout", topic,
-                                            peers = g.fanout.getOrDefault(topic).len
+    # https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md#mesh-maintenance
```

Comment:
@sinkingsugar in addition to this link, you should also review https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md#topic-membership, this covers the fanout peers that are missing from the maintenance snippet.

Reply:
@dryajov thanks, I will fix it, I committed to check it quick with NBC, still wip indeed, will add those bits

Reply:
Yes, the spec is a bit all over the place and it took me some time to realize that the mesh maintenance and the topic/mesh participation logic are doing the same thing, however, the snippet omits the fanout peers. But it makes perfect sense to think about it like this:

- Gossip peers are all the peers
- Fanout peers, are peers that we know are good peers, because we've published to those peers already
- When rebalancing the mesh
  - if we subscribing to a topic that is in fanout (we published there before), move it to the mesh table
  - if no enough fanout peers available, complete with peers from the gossip table

Also, IIRC Go/js do this as well...

Reply:
https://github.com/libp2p/go-libp2p-pubsub/blob/8945f91465bc3ba418688684a2c29c384cd79647/gossipsub.go#L1280

@dryajov Yes but completely changed now with 1.1
That's why I would not focus too much on this and leave the fanout excluded for now since 1.1 will come up soon.
Honestly this branch the more time stays in limbo the more 1.1 is delayed cos I have to rewrite a lot anyway

Reply:
I'll try to get this merged one way or another asap.